src/NLog/Internal/LogMessageTemplateFormatter.cs

```diff
@@ -0,0 +1,139 @@
+// 
+// * Neither the name of Jaroslaw Kowalski nor the names of its 
+//   contributors may be used to endorse or promote products derived from this
+//   software without specific prior written permission. 
+// 
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE 
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE 
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE 
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR 
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS 
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN 
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) 
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF 
+// THE POSSIBILITY OF SUCH DAMAGE.
+// 
+
+using System;
+using System.Globalization;
+using NLog.StructuredEvents;
+
+namespace NLog.Internal
+{
+    internal sealed class LogMessageTemplateFormatter : ILogMessageFormatter
+    {
+        public static readonly LogMessageTemplateFormatter DefaultAuto = new LogMessageTemplateFormatter(false);
+        public static readonly LogMessageTemplateFormatter Default = new LogMessageTemplateFormatter(true);
+        private static LogMessageTemplateFormatter DefaultFormatProvider { get; set; }
+        private static readonly Internal.StringBuilderPool _builderPool = new Internal.StringBuilderPool(Environment.ProcessorCount * 2);
+        private static readonly Func<System.Collections.Generic.IDictionary<object, object>> _dictionaryAllocator = () => { var properties = new Internal.PropertiesDictionary(); properties.HoleInsertMode = true; return properties; };
+
+        public IFormatProvider FormatProvider { get; private set; }
+        private readonly bool _forceTemplateRenderer;
+
+        private LogMessageTemplateFormatter(bool forceTemplateRenderer)
+        {
+            _forceTemplateRenderer = forceTemplateRenderer;
+        }
+
+        public bool HasProperties(LogEventInfo logEvent)
+        {
+            return !string.IsNullOrEmpty(logEvent.Message) && logEvent.Parameters != null && logEvent.Parameters.Length > 0;
+        }
+
+        public string FormatMessage(LogEventInfo logEvent)
+        {
+            if (HasProperties(logEvent))
+            {
+                // Prevent multiple layouts on different targets to render the same properties
+                lock (logEvent)
+                {
+                    using (var builder = _builderPool.Acquire())
+                    {
+                        var holeProperties = logEvent.HoleProperties;
+                        System.Collections.Generic.IDictionary<object, object> holeDictionary = holeProperties;
```

Comment:
Isn't this slower because of the interface calls? Or do you think the compiler optimizes this? 

(I think the interface is unneeded because `(Internal.PropertiesDictionary)holeDictionary` below)

Reply:
Code removed